nemo_aligner/models/mm/stable_diffusion/megatron_sd_draftp_model.py

```diff
@@ -0,0 +1,381 @@
+
+            time_range = np.flip(timesteps)
+            total_steps = timesteps.shape[0]
+
+            print(f"Running {sampler_draft_p.sampler.name} Sampling with {total_steps} timesteps")
+            iterator = tqdm(time_range, desc=f"{sampler_draft_p.sampler.name} Sampler", total=total_steps)
+
+            list_eps_draft_p = []
+            list_eps_init = []
+            truncation_steps = self.cfg.truncation_steps
+
+            denoise_step_kwargs = {
+                "unconditional_guidance_scale": self.unconditional_guidance_scale,
+                "unconditional_conditioning": u_cond,
+            }
+            for i, step in enumerate(iterator):
+
+                denoise_step_args = [total_steps, i, batch_size, device_draft_p, step, cond]
+
+                if i < total_steps - truncation_steps:
+                    with torch.no_grad():
+                        img_draft_p, pred_x0_draft_p, eps_t_draft_p = sampler_draft_p.single_ddim_denoise_step(
+                            prev_img_draft_p, *denoise_step_args, **denoise_step_kwargs
+                        )
+
+                        prev_img_draft_p = img_draft_p
+                else:
+
+                    img_draft_p, pred_x0_draft_p, eps_t_draft_p = sampler_draft_p.single_ddim_denoise_step(
+                        prev_img_draft_p, *denoise_step_args, **denoise_step_kwargs
+                    )
+                    list_eps_draft_p.append(eps_t_draft_p)
+
+                    with torch.no_grad():
+                        _, _, eps_t_init = sampler_init.single_ddim_denoise_step(
+                            prev_img_draft_p, *denoise_step_args, **denoise_step_kwargs
+                        )
+                        list_eps_init.append(eps_t_init)
+
+                    prev_img_draft_p = img_draft_p
+
+            last_states = [pred_x0_draft_p]
+
+            trajectories_predx0 = (
+                torch.stack(last_states, dim=0).transpose(0, 1).contiguous().view(-1, *last_states[0].shape[1:])
+            )
+            t_eps_draft_p = torch.stack(list_eps_draft_p).to(torch.device("cuda"))
+            t_eps_init = torch.stack(list_eps_init).to(torch.device("cuda"))
+
+            vae_decoder_output = []
+            for i in range(0, batch_size, self.vae_batch_size):
+                image = self.model.differentiable_decode_first_stage(trajectories_predx0[i : i + self.vae_batch_size])
+                vae_decoder_output.append(image)
+
+            vae_decoder_output = torch.cat(vae_decoder_output, dim=0)
+            vae_decoder_output = torch.clip((vae_decoder_output + 1) / 2, 0, 1) * 255.0
+
+            return vae_decoder_output, t_eps_draft_p, t_eps_init
+
+    def prepare_for_training(self):
+        configure_batch_sizes(
+            mbs=self.cfg.micro_batch_size,
+            gbs=self.cfg.global_batch_size,
+            dp=parallel_state.get_data_parallel_world_size(),
+        )
+        self.onload_adam_states()
```

Comment:
I think you don't need onload adam states since you never offload them, so you can delete